src/common/transformations/src/transformations/common_optimizations/remove_multi_subgraph_op_dangling_params.cpp

```diff
@@ -97,8 +146,47 @@ ov::pass::RemoveMultiSubGraphOpDanglingParams::RemoveMultiSubGraphOpDanglingPara
             }
             multi_subgraph_op->set_arguments(op_inputs);
         }
-        return false;
-    };
-    auto m = std::make_shared<pattern::Matcher>(multi_subgraph_op_pattern, matcher_name);
-    this->register_matcher(m, callback);
+        if (!outputs_to_remove.empty()) {
+            // we need to reconstruct operation with new number of outputs, we cannot reduce number of outputs of
+            // existing op
+            std::shared_ptr<ov::op::util::MultiSubGraphOp> new_op;
+            if (if_op) {
+                new_op = std::make_shared<opset10::If>(if_op->input_value(0));
+                new_op->set_arguments(multi_subgraph_op->input_values());
+            } else if (loop_op) {
+                auto new_loop_op = std::make_shared<opset10::Loop>(loop_op->input_value(0), loop_op->input_value(1));
+                new_loop_op->set_special_body_ports(loop_op->get_special_body_ports());
+                new_op = new_loop_op;
+                new_op->set_arguments(multi_subgraph_op->input_values());
+            } else if (ti_op) {
+                new_op = std::make_shared<opset10::TensorIterator>(multi_subgraph_op->input_values());
+            }
+            copy_runtime_info(multi_subgraph_op, new_op);
+            new_op->set_friendly_name(multi_subgraph_op->get_friendly_name());
+            for (int body_idx = 0; static_cast<size_t>(body_idx) < subgraphs_size; ++body_idx) {
+                new_op->set_function(body_idx, multi_subgraph_op->get_function(body_idx));
+                new_op->set_input_descriptions(body_idx, multi_subgraph_op->get_input_descriptions(body_idx));
+                new_op->set_output_descriptions(body_idx, new_op_out_desc.at(body_idx));
+            }
+            size_t removed_outs_counter = 0;
+            new_op->set_output_size(multi_subgraph_op->get_output_size() - outputs_to_remove.size());
+            for (size_t out_idx = 0; out_idx < multi_subgraph_op->get_output_size(); ++out_idx) {
+                if (outputs_to_remove.find(out_idx) != outputs_to_remove.end()) {
+                    ++removed_outs_counter;
+                    // Need to go through all output descriptors to reflect deleted output
+                    for (int body_idx = 0; static_cast<size_t>(body_idx) < subgraphs_size; ++body_idx) {
```

Comment:
I believe this is the same as for inputs, could we re-use this code?

Reply:
Unfortunately I don't see a way to reuse the lambda function effectively, I would create a new lambda for that